ethereum/p2p/src/main/java/tech/pegasys/pantheon/ethereum/p2p/upnp/UpnpNatManager.java

```diff
@@ -0,0 +1,487 @@
+
+    started = true;
+  }
+
+  /**
+   * Stop the manager. Must not be in stopped state.
+   *
+   * @throws IllegalStateException if stopped.
+   */
+  public synchronized void stop() {
+    if (!started) {
+      throw new IllegalStateException("Cannot stop already-stopped service");
+    }
+    upnpService.getRegistry().removeListener(registryListener);
+    upnpService.shutdown();
+
+    started = false;
+  }
+
+  /**
+   * Returns the first of the discovered services of the given type, if any.
+   *
+   * @param type is the type descriptor of the desired service
+   * @return the first instance of the given type, or null if none
+   */
+  private synchronized RemoteService getService(final String type) {
+    return recognizedServices.get(type);
+  }
+
+  /**
+   * Get the discovered WANIPConnection service, if any.
+   *
+   * @return the WANIPConnection Service if we have found it, or null.
+   */
+  public synchronized RemoteService getWANIPConnectionService() {
+    return getService(SERVICE_TYPE_WAN_IP_CONNECTION);
+  }
+
+  /**
+   * Get the local address on which we discovered our external IP address.
+   *
+   * <p>This can be useful to distinguish which network interface the external address was
+   * discovered on.
+   *
+   * @return the local address on which our GetExternalIP was discovered on, or null if no
+   *     GetExternalIP query has been performed successfully.
+   */
+  public String getDiscoveredOnLocalAddress() {
+    return this.discoveredOnLocalAddress;
+  }
+
+  /**
+   * Returns a CompletableFuture that will wait for the given service type to be discovered. No new
+   * query will be performed, and if the service has already been discovered, the future will
+   * complete in the very near future.
+   *
+   * @param serviceType is the service type to wait to be discovered.
+   * @return future that will return the desired service once it is discovered, or null if the
+   *     future is cancelled.
+   */
+  private CompletableFuture<RemoteService> discoverService(final String serviceType) {
+
+    return CompletableFuture.supplyAsync(
+        () -> {
+
+          // wait until our thread is interrupted (assume future was cancelled)
+          // or we discover the service
+          while (!Thread.currentThread().isInterrupted()) {
+            RemoteService service = getService(serviceType);
+            if (null != service) {
+              return service;
+            } else {
+              try {
+                Thread.sleep(100);
+              } catch (InterruptedException e) {
+                // fall back through to "isInterrupted() check"
+              }
+            }
+          }
+          return null;
+        });
+  }
+
+  /**
+   * Sends a UPnP request to the discovered IGD for the external ip address.
+   *
+   * @return A CompletableFuture that can be used to query the result (or error).
+   */
+  public synchronized CompletableFuture<String> queryExternalIPAddress() {
+    if (!started) {
```

Comment:
Should all of these public methods be similarly validated against the `started` status?